sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,251 @@
+// Package listener implements a high-level interface to the apparmor
+// notification mechanism. It can be used to build userspace applications
+// which respond to apparmor prompting profiles.
+package listener
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/osutil/epoll"
+	"github.com/snapcore/snapd/sandbox/apparmor/notify"
+)
+
+// Listener contains low-level components for receiving notification requests
+// and responding with notification responses.
+type Listener struct {
+	// R is a channel with incoming requests. Each request is asynchronous
+	// and needs to be replied to.
+	R chan *Request
+	// E is a channel for receiving asynchronous error messages from
+	// concurrently running parts of the listener system.
+	E chan error
+
+	notifyFile *os.File
+	poll       *epoll.Epoll
+}
+
+// Request is a high-level representation of an apparmor prompting message.
+//
+// Each request must be replied to by writing a boolean to the YesNo channel.
+type Request struct {
+	l *Listener
+
+	// Pid is the identifier of the process triggering the request.
+	Pid uint32
+	// Label is the apparmor label on the process triggering the request.
+	Label string
+	// SubjectUID is the UID of the subject that triggered the prompt
+	SubjectUid uint32
+
+	// Path is the path of the file, as seen by the process triggering the request.
+	Path string
+	// Permission is the opaque permission that is being requested.
+	Permission interface{}
+	// YesNo is a channel for writing the response.
+	YesNo chan bool
+}
+
+func newRequest(l *Listener, msg *notify.MsgNotificationFile) *Request {
+	var perm interface{}
+	if msg.Class == notify.AA_CLASS_FILE {
+		_, deny, _ := msg.DecodeFilePermissions()
```

Comment:
The only error for `DecodeFilePermissions()` is if `msg.Class != AA_CLASS_FILE`, which we check immediately before, but we may as well add the check here just in case we end up using a similar pattern in the future, and certainly if we introduce other types of message classes.